### DISCUSSION TITLE
Who is responsible for sending an IP packet through a TUN adapter in a point-to-point tunnel? (kernel or programmer/code)?

### DISCUSSION THREAD

**Discussion by M0dEx** [quincy-rs/quincy#55] [Q&A] [answered]
Hello.
The question is a little off Quincy :)
I want to develop my own solution for organizing a point-to-point VPN tunnel.
Using quincy as an example, a tun adapter is created on Linux and the necessary routes are registered in the routing table.
This allows you to wrap any traffic (TCP/UDP/ICMP) through tun.
The tunnel itself can be on WSS (WebSockets Secure) or on QUIC.

I want to understand one point: who is responsible for working with the IP packet when sending it through the created WSS/QUIC tunnel - the Linux kernel or is it implemented by the programmer in the client/server?

Originally asked by @vukituso

**Comment by M0dEx** [comment, marked as answer]
The overall packet pipeline goes like this:

Application sends packet -> kernel routes to the the correct interface -> TUN interface -> tunneling software receives the packet from the TUN interface file descriptor -> tunneling software sends the encapsulated packet through an open socket as if it was standard application traffic -> kernel routes to the correct interface -> interface sends it over a network.

Sorry for formatting, I am on mobile right now.

**Reply by vukitoso** [reply]
Got it, thank you very much.
